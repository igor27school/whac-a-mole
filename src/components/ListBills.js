import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { fetchBillsFromServer } from '../actions/ActionCreators'
import BillSummary from './BillSummary'

/**
* @description Displays bills as a list.
*/
export class ListBills extends Component {
  static propTypes = {
    hasBills: PropTypes.bool.isRequired,
    bills: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
      })
    ).isRequired,
  }
  componentDidMount() {
    const { hasBills, fetchBillsFromServer } = this.props
    if (!hasBills) {
      fetchBillsFromServer()
    }
  }
  render() {
    const { bills } = this.props
    return (
      <div>
        <h2>BILLS</h2>
        <ul>
          {bills.map(bill => (
            <li key={bill.id}>
              <BillSummary billId={bill.id}/>
            </li>
          ))}
        </ul>
      </div>
    )
  }
}

function mapStateToProps ({ bills }) {
  return {
    bills: bills.allIds.map(bill => bills.byId[bill]),
    hasBills: bills.allIds.length > 0,
  }
}

function mapDispatchToProps (dispatch) {
  return {
    fetchBillsFromServer: () => dispatch(fetchBillsFromServer()),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ListBills)
