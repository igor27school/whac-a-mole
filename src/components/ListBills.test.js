import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { ListBills } from './ListBills'

jest.mock('../utils/ServerAPI')

describe('ListBills tests', () => {
  const setup = () => {
    const props = {
      hasBills: true,
      bills: [
        {
          id: 'bill1',
          title: 'bill title 1',
        },
        {
          id: 'bill2',
          title: 'bill title 2',
        },
      ],
      fetchBillsFromServer: jest.fn(),
    }
    const renderer = new ShallowRenderer()
    return {
      props,
      renderer
    }
  }

  it('Bills match the snapshot', () => {
    const { renderer, props } = setup()
    renderer.render(<ListBills {...props}/>)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
