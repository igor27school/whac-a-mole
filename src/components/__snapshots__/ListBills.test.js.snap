// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBills tests Bills match the snapshot 1`] = `
<div>
  <h2>
    BILLS
  </h2>
  <ul>
    <li>
      bill title 1
    </li>
    <li>
      bill title 2
    </li>
  </ul>
</div>
`;
